lib/models/calculator.js

```diff
@@ -0,0 +1,191 @@
+/**
+ * The CQL calculator. This calls the cql-execution framework and formats the results as neccesary.
+ */
+const _ = require('lodash');
+const CqmModels = require('cqm-models');
+const cql = require('cqm-models').CQL;
+const ResultsHelpers = require('../helpers/results_helpers');
+const CalculatorHelpers = require('../helpers/calculator_helpers');
+
+module.exports = class Calculator {
+  /**
+   * Generate calculation results for patients against a CQL measure.
+   *
+   * @param {Measure} measure - The measure population to calculate on.
+   * @param {PatientSource} patientSource - The patientSource to run calculations on.
+   * @param {Hash} valueSetsByOid - all ValueSets relevant to the measure.
+   * @param {Hash} options - contains options for measure calculation.
+   */
+  static calculate(measure, patientSource, valueSetsByOid, options = {}) {
+    // We store both the calculation result and the calculation code based on keys derived from the arguments
+    const resultsByPatient = {};
+
+    // Grab start and end of Measurement Period
+    let start;
+    let end;
+    // Override default measure_period with effective_date if available
+    if (options && options.effective_date != null) {
+      start = CalculatorHelpers.getConvertedTime(options.effective_date);
+      if (options.effective_date_end != null) {
+        end = CalculatorHelpers.getConvertedTime(options.effective_date_end);
+      } else {
+        end = CalculatorHelpers.getConvertedTimeEndOfYear(options.effective_date);
+      }
+    } else {
+      start = CalculatorHelpers.getConvertedTime(measure.measure_period.low.value);
+      end = CalculatorHelpers.getConvertedTime(measure.measure_period.high.value);
+    }
+
+    const startCql = cql.DateTime.fromDate(start, 0); // No timezone offset for start
+    const endCql = cql.DateTime.fromDate(end, 0); // No timezone offset for stop
+
+    // Construct CQL params
+    const params = { 'Measurement Period': new cql.Interval(startCql, endCql) };
+
+    // Grab execution date time from options
+    let executionDateOption;
+    // Override default measure_period with effective_date if available else default to new Date()
+    if (options && options.execution_date != null) {
+      executionDateOption = CalculatorHelpers.getConvertedTime(options.effective_date);
+    }else{
+      executionDateOption = new Date();
+    }
+
+    // Create the execution DateTime that we pass into the engine
+    const executionDateTime = cql.DateTime.fromDate(executionDateOption, '0');
```

Comment:
For some reason my comment about this is hidden. We should have this be an option that can be passed in. If nothing is passed in, then set it to this line.